Add dedicated "Usage" page to documentation and revise start page (#114)
martius-lab/cluster_utils: docs/usage.rst

*****
Usage
*****

> Run Hyperparameter Search

Reviewer: ```suggestion
Run Batch of Jobs
```

Alternatively, "Run Array of Jobs" (more fitting to the cluster terminology, but probably less familiar for people)

=========================

> cluster_utils provides two main commands to run hyperparameter search in different
> modes:

Reviewer: ```suggestion
cluster_utils provides two main commands to interface with the cluster:
```

In my view, "hyperparameter search" is not the primary description of what cluster_utils does. It is the major usecase, but fundamentally I see cluster_utils as providing a convenient and unified interface to the cluster.

Author: I was actually a bit unsure if "hyperparameter search" is a good overarching term, so I'm happy to change it.  Just "interface with the cluster" seems a bit too vague to me, though.  How about "to run batches of jobs on the cluster" (picking up the nomenclature from the headline)?

> - ``grid_search``:  Simple grid search over specified hyperparameter ranges.

Reviewer: ```suggestion
- ``grid_search``:  Simple grid search over specified parameter ranges.
```

- ``hp_optimization``:  Uses sampling-based optimization to search for best combination
  of hyperparameters within the specified ranges.

They are both run as modules via ``python -m`` and expect a configuration file as
argument:

.. code-block:: bash

   python3 -m cluster_utils.grid_search config_for_grid_search.json

   python3 -m cluster_utils.hp_optimization config_for_hp_optim.json

See :doc:`configuration` for information on the expected structure of the config file
(note that there are differences between the two methods!).

Optionally, a list of key-value arguments can be provided in addition, to overwrite
single settings from the config file.  Use dot-notation to specify nested keys.  Example:

.. code-block:: bash

   python3 -m cluster_utils.grid_search config.json \

>      'results_dir="/tmp"' 'optimization_setting.run_local=True'

Reviewer: Is `optimization_setting.run_local` a setting we actually support (if it is, I missed that)? We should only list commands that directly work when copied.

Author: Yes, that is an actual setting.  It suppresses the "Do you want to run locally?" prompt when running locally (see [docs](https://martius-lab.github.io/cluster_utils/configuration.html#confval-optimization_setting.run_local).
However, I now realised that this is for some reason an hp_optimization-specific setting, i.e. it doesn't work with grid_search.  I'll change the example accordingly.

Both commands can also be run with ``--help`` to get a complete list of arguments.

> You can abort the hyperparameter search with Ctrl + C at any time.

Reviewer: ```suggestion
You can abort `cluster_utils` with Ctrl + C at any time. All running jobs are stopped, and submitted jobs are removed from the cluster queue.
```

   list_idle_jobs, show_job, stop_remaining_jobs
   >>>

You now may enter one of the listed commands or simply press Enter to leave the prompt
without executing a command.

.. important::

   While the prompt is open, the main loop is blocked, i.e. no new jobs will be
   submitted during that time.

Commands
--------

.. I'm a bit misusing the confval directive here, but I think as long as there is no
   name collision with an actual config value, this should be fine and much easier than
   adding a dedicated directive.

.. confval:: list_jobs

   List IDs of all jobs that have been submitted so far (including finished ones).

.. confval:: list_running_jobs

   List IDs of all jobs that are currently running.

.. confval:: list_successful_jobs

   List IDs of all jobs that finished successfully.

.. confval:: list_idle_jobs

   List IDs of all jobs that have been submitted but not yet started.

.. confval:: show_job

   Will ask for a job ID and show information about this job.

.. confval:: stop_remaining_jobs

   Abort all currently running jobs as well as jobs that already have been submitted but
   didn't start yet.

>    This will not stop submission of new jobs.  If you want to stop cluster_utils

Reviewer: It does not stop cluster_utils fully?  Is this the case for hp_optimization? If so, this is a bug, because on grid_search, it does (and is the intended outcome of the command).

Author: Yes, this is the case for hp_optimization.  I think the difference to grid_search is that there all jobs will be submitted at once while hp_optimization waits for first jobs to finish before submitting the next ones.
From what I understood, the command only cancels running and submitted jobs but doesn't exit the main loop, so once they are cancelled, it proceeds submitting the next batch.  I have to admit, I was a bit confused about this behaviour back when I noticed it :D.

Another funny thing I just noticed: After calling it, I get a rather high number of failed jobs displayed in the "started execution" progress bar which then goes *down* over the next few seconds.  So also something is wrongly counted there.

Anyway, if the purpose is to stop everything, do we actually need that command?  Does it have any advantage over stopping with Ctrl+C?

Author: I created issue #116 for this.
